File: Data Structures & Algorithms/two-integer-sum/submission-0.java
class Solution {
    public int[] twoSum(int[] nums, int target) {
        HashMap<Integer, Integer> hashMap = new HashMap<>();
        int difference;
        int[] returnVals = new int[2];


        for(int i = 0; i < nums.length; i++) {
            difference = target - nums[i];
            hashMap.put(difference, i);
        }
        
        for(int j = 0; j < nums.length; j++) {
            if(hashMap.containsKey(nums[j])) {
                if(j < hashMap.get(nums[j])) {
                    returnVals[0] = j;
                    returnVals[1] = hashMap.get(nums[j]);
                    return returnVals;
                } else if (j > hashMap.get(nums[j])) {
                    returnVals[0] = hashMap.get(nums[j]);
                    returnVals[1] = j;
                    return returnVals;
                }
            }
            
        }
        return null;
        
    }
}

// if(hashMap.containsKey(difference)) {
//                 System.out.println("in loop");
//                 if(hashMap.get(difference) < j){
//                     returnVals[0] = hashMap.get(difference);
//                     returnVals[1] = j;
//                     return returnVals;
//                 } else if (hashMap.get(difference) > j){
//                     returnVals[1] = hashMap.get(difference);
//                     returnVals[0] = j;
//                     return returnVals;
//                 }
//                 System.out.println(hashMap.get(difference));
//             }

File: Data Structures & Algorithms/binary-search/submission-0.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        lo, hi = 0, len(nums) - 1

        while lo <= hi:
            mid = (hi + lo) // 2

            if target < nums[mid]:
                hi = mid - 1
            elif target > nums[mid]:
                lo = mid + 1
            else: 
                return mid

        return -1



        # hi, lo, mid = len(nums) - 1, 0, math.floor((len(nums) - 1)/2)
        # print(hi, mid, lo)
        # if target == nums[mid]:
        #     return mid
        # elif lo == mid:
        #     if nums[hi] == target:
        #         return hi
        #     else:
        #         return -1
        # else:
        #     if target > mid:
        #         return self.search(nums[mid:hi+1], target)
        #     else:
        #         return self.search(nums[lo:mid], target)
